### ISSUE TITLE
Add a ser_json_datetime config option for configuring how datetimes are serialized.

### ISSUE THREAD

**Issue by ollz272** [pydantic/pydantic#10454]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

We have a use case where we want to have datetime objects in our pydantic models, but when serialized they become unix timestamps/ millisecond unix timestamps. Heres a rough way that we currently do it:

```python
from typing import Annotated
import datetime as dt
from pydantic import BaseModel, PlainSerializer


T_SerializedDatetime = Annotated[
    dt.datetime,
    PlainSerializer(
        lambda x: x.timestamp() * 1000,
        return_type=int,
        when_used="json",
    ),
]

class Value(BaseModel):
    series: dict[T_SerializedDatetime, float]
```

This works fine, however we have noticed that when we're serializing a large data series we see a lot of time spent computing the `x.timestamp() * 1000` value within python, which leads to some performance issues.

What would be really great is if we could do the following:

```python
class Value(BaseModel):
    series: dict[dt.datetime, float]


    model_config = ConfigDict(ser_json_datetime="milliseconds_float")
```

And in this case pydantic would do the computation of the millisecond timestamp/second timestamp within rust, thus increasing performance for us.

I would invision the modes being similar to the current state of `ser_json_timedelta`, being able to take `iso8601, seconds_float, milliseconds_float` as options, and in order to keep existing behaviour defaulting to `iso8601`.

Is this something that the pydantic team would support having as a feature?

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [X] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [X] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by ollz272** [comment]
As a note, I'm happy to work on this feature, though may need a bit of guidance in this (i know bits and bobs now from the work on `ser_json_timedelta`, but as this would be a new config option im guessing theres more to consider!!)

**Comment by davidhewitt** [comment]
Some kind of option on this line seems very reasonable to me, thanks!

FWIW we currently infer the seconds / microseconds split for timestamps - see https://github.com/pydantic/pydantic/issues/7940 and https://github.com/pydantic/speedate/issues/14

I think any implementation / design here strongly wants that complementary problem solved so that you don't fall into edge cases when deserializing. Perhaps there's room for a pair of options `ser_json_datetime='float'` and then a config mode for `timestamp_unit='milliseconds'` (with the config affecting the unit for both directions)? Open to opinions🤔

**Comment by sydney-runkle** [comment]
Hmm, given that we just added `milliseconds_float`, `seconds_float` options for `ser_json_timedelta`, perhaps we should just use a consistent pattern here?

**Comment by ollz272** [comment]
Thanks, glad its something worth considering!

I think i'd agree with the milliseconds_float, seconds_float (probably int in this case?) to keep a consistent pattern

**Comment by ollz272** [comment]
I've done a very rough implementation in a draft PR for pydantic core which uses this method anyway so somewhat bias, but happy to change depending on what you guys feel

**Comment by ollz272** [comment]
I've set the PR i have done with this implementation as ready to review, but welcome any design feedback if we want to change this 👍🏻

**Comment by sydney-runkle** [comment]
> I've set the PR i have done with this implementation as ready to review, but welcome any design feedback if we want to change this 👍🏻

Awesome! Will review tomorrow :)

**Comment by ollz272** [comment]
Hi @sydney-runkle , did you get a chance to have a look?

**Comment by sydney-runkle** [comment]
Alright, so we've discussed a bit of a switch up here re the API. This piggy backs off of @davidhewitt's [comment](https://github.com/pydantic/pydantic/issues/10454#issuecomment-2368636308) above.

The new API we're proposing

Config settings:
* `val_timestamp_unit`: `'seconds'`, `'milliseconds'`, or `'infer'`, defaults to `'infer'`
* `ser_timestamp_unit`: `'seconds'`, `'milliseconds'`, or `'infer'`, defaults to `'infer'`
* `ser_json_timedelta` = `'float'` or `'iso8601'`
* `ser_json_datetime` = `'float'` or `'iso8601'`

Unfortunately, this will make it such that some of the work from your previous PR is reverted 😞, at least in terms of config setting changes. We can likely reuse a lot of the internal logic, and likewise for your current `pydantic-core` PR.

This change will also make it such that we resolve https://github.com/pydantic/pydantic/issues/7940. You can see some related work already done here: https://github.com/pydantic/pydantic-core/pull/1052.

Additionally, this should fix some issues re round trips for cases around our current watershed.

Apologies for this change up after you've done some of the work to implement support. We would of course love your continued help with development here, if you're still interested.

Would love to hear your thoughts on this API given that you've requested support for customization like the above!

**Comment by ollz272** [comment]
@sydney-runkle  Makes sense, i may be a little constrained for time soon but i'm hoping I can get this through for you guys for the 2.10 release (Or at least, I would love to see it in that release!!)

I may have some follow up questions on this once, i dive into the details. I'll work on the pydantic core side first ofc. Do we envisage any changes needed to speeddate?

**Comment by sydney-runkle** [comment]
@ollz272,

Great! We're super excited to have you contributing :). Looking forward to some more collaboration.

**Comment by sydney-runkle** [comment]
Hey @ollz272, hope you're having a good weekend. Any estimate as to when you might be able to pick this up?

**Comment by ollz272** [comment]
@sydney-runkle apologies, i have a feeling that with new work and life commitments im not going to be able to get round to this any time soon

**Comment by sydney-runkle** [comment]
@ollz272,

No worries. I'll pick up these changes. Thanks for your help on this front - definitely kickstarted our movement to a better API + more extensive functionality.

**Comment by CharlesPerrotMinotHCHB** [comment]
Being able to pass a custom format would be really useful as well.
For example, we do serialize all datetimes with `dt.isoformat(timespec="seconds")`
